docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,212 @@
+- Ensure the design allows for future extensions, such as replace-by-priority and
+  allowing multiple pending transactions per sender, to be incorporated easily.
+
+Note, not all of these properties will be addressed by the proposed changes in
+this ADR. However, this proposal will ensure that any unaddressed properties
+can be addressed in an easy and extensible manner in the future.
+
+## Alternative Approaches
+
+When considering which approach to take for a priority-based flexible and
+performant mempool, there are two core candidates. The first candidate in less
+invasive in the required  set of protocol and implementation changes, which
+simply extends the existing `CheckTx` ABCI method. The second candidate essentially
+involves the introduction of new ABCI method(s) and would require a higher degree
+of complexity in protocol and implementation changes, some of which may either
+overlap or conflict with the upcoming introduction of [ABCI++](https://github.com/tendermint/spec/blob/master/rfc/004-abci%2B%2B.md).
+
+For more information on the various approaches and proposals, please see the
+[mempool discussion](https://github.com/tendermint/tendermint/discussions/6295).
+
+## Decision
+
+To incorporate a priority-based flexible and performant mempool in Tendermint Core,
+we will introduce new fields, `priority` and `sender`, into the `ResponseCheckTx`
+type and augment the existing mempool data structure to facilitate prioritization
+of uncommitted transactions in addition to extended functionality such as
+replace-by-priority and allowing multiple transactions to exist from the same
+sender with varying priorities.
+
+## Detailed Design
+
+### CheckTx
+
+We introduce the following new fields into the `ResponseCheckTx` type:
+
+```diff
+message ResponseCheckTx {
+  uint32         code       = 1;
+  bytes          data       = 2;
+  string         log        = 3;  // nondeterministic
+  string         info       = 4;  // nondeterministic
+  int64          gas_wanted = 5 [json_name = "gas_wanted"];
+  int64          gas_used   = 6 [json_name = "gas_used"];
+  repeated Event events     = 7 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string         codespace  = 8;
++ int64          priority   = 9;
++ string         sender     = 10;
+}
+```
+
+It is entirely up the application in determining how these fields are populated
+and with what values, e.g. the `sender` could be the signer and fee payer 
+the transaction, the `priority` could be the cumulative sum of the fee(s).
+
+Only `sender` is required, while `priority` can be omitted which would result in
+using the default value of zero.
+
+### Mempool
+
+The existing concurrent-safe linked-list will be removed entirely in favor of a
+thread-safe map of `<sender:*Tx>`, i.e a mapping from `sender` to a single `*Tx`
+object, where each `*Tx` is the next valid and processable transaction from the
+given `sender`.
```

Comment:
Can there not be more than one tx by the same sender in a single block?

Reply:
No, any given sender can only have one transaction in the mempool at a time (for now). `sender` is just a one-dimensional grouping that can later be extended. We need some sort of bucketing/dimensionality to get started with and sender is the most applicable across many projects and apps.

Reply:
Thanks for the clarification. Would it be okay to make this a little more explicit. So in my head I have something like this:

```go
type mempool struct {
   txs map[sender]*Tx 
   priorityQueue // min-max heap based on priority where the values are what? the sender? the tx themselves?
   // other stuff
}
```
Is this somewhat right?

Reply:
Something like that, yeah.

Reply:
Is the value within the priority queue the sender string?

Reply:
No, the priority queue index contains references to a `*Tx`, prioritized by `Tx.Priority `.

Reply:
Right. Basically the number of txs per sender in the mempool should be a parameter, but we discussed starting with just one to get the rest of the design in place and then add the ability for multiple tx per sender up to some limit.

It's also possible in many cases instead of sending multiple txs to just send a single multi-msg tx, but this assumes the app has multi-msg support and you can sign a large enough tx and so on.